src/bson_mapper/test/bson_streambuf.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "catch.hpp"
+
+#include <bson_mapper/bson_streambuf.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/types.hpp>
+
+using namespace bson_mapper;
+
+/**
+ * A callable object that compares a given BSON document to an original reference.
+ * This is used for unit testing.
+ */
+class doc_validator_callback {
+   public:
+    doc_validator_callback(bsoncxx::document::value doc) : _doc(doc) {
+    }
+
+    /**
+     * This asserts that the given document is the same as the reference document.
+     */
+    void operator()(bsoncxx::document::value v) {
+        REQUIRE(v.view() == _doc.view());
```

Comment:
This looks good! Can you add a counter to this class that increments every time this is called? Then at the end (after line 62), `REQUIRE(cb.count() == 2)`. That way, we can see that this is happening both times.